src/webgpu/util/f32_interval.ts

```diff
@@ -0,0 +1,213 @@
+import { assert } from '../../common/util/util.js';
+
+import { kValue } from './constants.js';
+import { flushSubnormalNumber, nextAfter, oneULP } from './math.js';
+
+/** Represents a closed interval in the f32 range */
+export class F32Interval {
+  public readonly begin: number;
+  public readonly end: number;
+  private static _infinite: F32Interval;
+
+  /** Constructor
+   * @param begin non-NaN number indicating the lower bound of the interval
+   * @param end non-NaN number indicating the upper bound of the interval
+   */
+  public constructor(begin: number, end: number) {
+    assert(begin <= end, `begin (${begin}) must be equal or before end (${end})`);
+    assert(!Number.isNaN(begin), `begin must not be NaN`);
+    assert(!Number.isNaN(end), `end must not be NaN`);
+
+    if (begin === Number.NEGATIVE_INFINITY || begin < kValue.f32.negative.min) {
+      this.begin = Number.NEGATIVE_INFINITY;
+    } else if (begin === Number.POSITIVE_INFINITY || begin > kValue.f32.positive.max) {
+      this.begin = kValue.f32.positive.max;
+    } else {
+      this.begin = begin;
+    }
+
+    if (end === Number.POSITIVE_INFINITY || end > kValue.f32.positive.max) {
+      this.end = Number.POSITIVE_INFINITY;
+    } else if (end === Number.NEGATIVE_INFINITY || end < kValue.f32.negative.min) {
+      this.end = kValue.f32.negative.min;
+    } else {
+      this.end = end;
+    }
+  }
+
+  /** Is a point or interval completely within this interval? */
+  public contains(x: number | F32Interval): boolean {
+    if (Number.isNaN(x)) {
+      // Being the infinite interval indicates that the accuracy is not defined for this test, so the test is just
+      // checking that this input doesn't cause the implementation to misbehave, so NaN is acceptable.
+      return this.begin === Number.NEGATIVE_INFINITY && this.end === Number.POSITIVE_INFINITY;
+    }
+    const i = toInterval(x);
+    return this.begin <= i.begin && this.end >= i.end;
+  }
+
+  /** Create interval with the tightest bounds that includes all provided intervals */
+  static span(...intervals: F32Interval[]): F32Interval {
+    assert(intervals.length > 0, `span of an empty list of F32Intervals is not allowed`);
+    const mins = new Set<number>();
```

Comment:
nit: Using sets here is seems unnecessarily expensive. Can't you just loop over the intervals and record the smallest of each interval's `.begin` and largest of each interval's `.end` ?

Reply:
I wasn't suggesting brute-force testing all permutations as a replacement for Sets - removing duplicates is important for the reasons you gave. Just here, in `span`, which looks like very hot code, a single loop over the intervals and tracking min / max is clearly going to be faster than any implementation using sets.

Your latest implementation is exactly what I had imagined - looks great!

Reply:
Sorry about this, I confounded two different areas I was working on and was responding to something else I had been thinking about vs what you had actually asked.